README.md
<h1 align="center">Weight Slider Widget for Flutter</h1>
<p align="center">
  <img src="https://api.codemagic.io/apps/5e1b33e348f5bc0b2ae8d4c4/5e1b33e348f5bc0b2ae8d4c3/status_badge.svg" />
  <a href="https://pub.dev/packages/weight_slider">
    <img alt="Pub Version" src="https://img.shields.io/pub/v/weight_slider">
  </a>
  <a href="https://raw.githubusercontent.com/coval-solutions/weight-slider/master/LICENSE">
    <img alt="License: MIT" src="https://img.shields.io/badge/license-MIT-yellow.svg" target="_blank" />
  </a>
</p>

## ✨ Demo
<p align="center"><img width="20%" src="https://raw.githubusercontent.com/coval-solutions/weight-slider/master/images/1.gif" alt="weight slider"/></p>

## 🚀 Usage
```dart
class _MyHomePageState extends State<MyHomePage> {
  int weight = 70;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Padding(
          padding: EdgeInsets.all(64),
          child: WeightSlider(
            weight: weight,
            minWeight: 40,
            maxWeight: 120,
            onChange: (val) => setState(() => this.weight = val),
            unit: 'kg', // optional
          ),
        ),
      ),
    );
  }
}
```

## 📝 License
This project is [MIT](https://raw.githubusercontent.com/coval-solutions/weight-slider/master/LICENSE) licensed.

CHANGELOG.md
# Changelog

## 1.3.0-nullsafety+1

* Initial release with null safety

## 1.2.2

  * Replaced `width` property with a changeable `height` property, you can adjust the width via outer padding

## 1.2.1

  * Upgrade the `flutter_svg` dependency from, `^0.19.1` to `^0.21.0+1`
  * Removed padding from `WeightSlider`, allowing for flexibility
  * Updated `example` to use the removed padding

## 1.2.0

  * Upgrade the `flutter_svg` dependency from, `^0.17.4` to `^0.19.1`
  * Added example app
  * Added `unit` parameter and display alongside the current value (default: `kg`)

## 1.1.0

  * Added Flutter dependency requirements, `>=1.6.7 <2.0.0`
  * Updated Dart SDK requirements, from `>=2.1.0 <3.00` to `>=2.7.0 <3.0.0`
  * Upgrade the `flutter_svg` dependency from, `^0.15.0` to `^0.17.4`

## 1.0.2

  * Arrow slider indicator colour is now the primary colour of the app theme
  * Renamed onChanged parameter to onChange

## 1.0.1

  * Ran `flutter format`

## 1.0.0

  * Created weight slider widget

lib/weight_slider.dart
library weight_slider;

import 'package:flutter/material.dart';
import 'package:weight_slider/src/weight_slider_background.dart';
import 'package:weight_slider/src/weight_slider_internal.dart';

class WeightSlider extends StatelessWidget {
  final int weight;
  final int minWeight;
  final int maxWeight;
  final String unit;
  final double height;
  final ValueChanged<int> onChange;

  const WeightSlider(
      {Key? key,
      this.weight = 80,
      this.minWeight = 30,
      this.maxWeight = 130,
      this.unit = 'kg',
      this.height = 100,
      required this.onChange})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return WeightBackground(
      height: this.height,
      child: LayoutBuilder(
        builder: (context, constraints) {
          return constraints.isTight
              ? Container()
              : WeightSliderInternal(
                  minValue: this.minWeight,
                  maxValue: this.maxWeight,
                  value: this.weight,
                  unit: this.unit,
                  onChange: this.onChange,
                  width: constraints.maxWidth,
                );
        },
      ),
    );
  }
}


lib/src/weight_slider_interal.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class WeightSliderInternal extends StatelessWidget {
  WeightSliderInternal({
    Key key,
    @required this.minValue,
    @required this.maxValue,
    @required this.value,
    @required this.onChange,
    @required this.width,
  })  : scrollController = new ScrollController(
          initialScrollOffset: (value - minValue) * width / 3,
        ),
        super(key: key);

  final int minValue;
  final int maxValue;
  final int value;
  final ValueChanged<int> onChange;
  final double width;
  final ScrollController scrollController;

  double get itemExtent => width / 3;

  int _indexToValue(int index) => minValue + (index - 1);

  @override
  build(BuildContext context) {
    int itemCount = (maxValue - minValue) + 3;
    return NotificationListener(
      onNotification: _onNotification,
      child: new ListView.builder(
        controller: scrollController,
        scrollDirection: Axis.horizontal,
        itemExtent: itemExtent,
        itemCount: itemCount,
        physics: BouncingScrollPhysics(),
        itemBuilder: (BuildContext context, int index) {
          int itemValue = _indexToValue(index);
          bool isExtra = index == 0 || index == itemCount - 1;

          return isExtra
              ? new Container()
              : GestureDetector(
                  behavior: HitTestBehavior.translucent,
                  onTap: () => _animateTo(itemValue, durationMillis: 50),
                  child: FittedBox(
                    child: Text(
                      itemValue.toString(),
                      style: _getTextStyle(context, itemValue),
                    ),
                    fit: BoxFit.scaleDown,
                  ),
                );
        },
      ),
    );
  }

  TextStyle _getDefaultTextStyle() {
    return new TextStyle(
      color: Color.fromRGBO(196, 197, 203, 1.0),
      fontSize: 14.0,
    );
  }

  TextStyle _getHighlightTextStyle(BuildContext context) {
    return new TextStyle(
      color: Theme.of(context).primaryColor,
      fontSize: 28.0,
    );
  }

  TextStyle _getTextStyle(BuildContext context, int itemValue) {
    return itemValue == value
        ? _getHighlightTextStyle(context)
        : _getDefaultTextStyle();
  }

  bool _userStoppedScrolling(Notification notification) {
    return notification is UserScrollNotification &&
        notification.direction == ScrollDirection.idle &&
        scrollController.position.activity is! HoldScrollActivity;
  }

  _animateTo(int valueToSelect, {int durationMillis = 200}) {
    double targetExtent = (valueToSelect - minValue) * itemExtent;
    scrollController.animateTo(
      targetExtent,
      duration: new Duration(milliseconds: durationMillis),
      curve: Curves.decelerate,
    );
  }

  int _offsetToMiddleIndex(double offset) => (offset + width / 2) ~/ itemExtent;

  int _offsetToMiddleValue(double offset) {
    int indexOfMiddleElement = _offsetToMiddleIndex(offset);
    int middleValue = _indexToValue(indexOfMiddleElement);
    middleValue = math.max(minValue, math.min(maxValue, middleValue));

    return middleValue;
  }

  bool _onNotification(Notification notification) {
    if (notification is ScrollNotification) {
      int middleValue = _offsetToMiddleValue(notification.metrics.pixels);
      if (_userStoppedScrolling(notification)) {
        _animateTo(middleValue);
      }

      if (middleValue != value && this.onChange != null) {
        this.onChange(middleValue);
      }
    }

    return true;
  }
}


lib/src/weight_slider_internal.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class WeightSliderInternal extends StatelessWidget {
  final int minValue;
  final int maxValue;
  final int value;
  final String unit;
  final ValueChanged<int> onChange;
  final double width;
  final ScrollController scrollController;

  WeightSliderInternal({
    Key? key,
    required this.minValue,
    required this.maxValue,
    required this.value,
    required this.unit,
    required this.onChange,
    required this.width,
  })  : scrollController = new ScrollController(
          initialScrollOffset: (value - minValue) * width / 3,
        ),
        super(key: key);

  double get itemExtent => width / 3;

  int _indexToValue(int index) => minValue + (index - 1);

  @override
  build(BuildContext context) {
    int itemCount = (maxValue - minValue) + 3;
    return NotificationListener(
      onNotification: _onNotification,
      child: new ListView.builder(
        controller: scrollController,
        scrollDirection: Axis.horizontal,
        itemExtent: itemExtent,
        itemCount: itemCount,
        physics: BouncingScrollPhysics(),
        itemBuilder: (BuildContext context, int index) {
          int itemValue = _indexToValue(index);
          bool isExtra = index == 0 || index == itemCount - 1;

          return isExtra
              ? new Container()
              : GestureDetector(
                  behavior: HitTestBehavior.translucent,
                  onTap: () => _animateTo(itemValue, durationMillis: 50),
                  child: FittedBox(
                    child: Text(
                      itemValue != value
                          ? itemValue.toString()
                          : itemValue.toString() + this.unit,
                      style: _getTextStyle(context, itemValue),
                    ),
                    fit: BoxFit.scaleDown,
                  ),
                );
        },
      ),
    );
  }

  TextStyle _getDefaultTextStyle() {
    return new TextStyle(
      color: Color.fromRGBO(196, 197, 203, 1.0),
      fontSize: 14.0,
    );
  }

  TextStyle _getHighlightTextStyle(BuildContext context) {
    return new TextStyle(
      color: Theme.of(context).primaryColor,
      fontSize: 28.0,
    );
  }

  TextStyle _getTextStyle(BuildContext context, int itemValue) {
    return itemValue == value
        ? _getHighlightTextStyle(context)
        : _getDefaultTextStyle();
  }

  bool _userStoppedScrolling(Notification notification) {
    return notification is UserScrollNotification &&
        notification.direction == ScrollDirection.idle &&
        scrollController.position.activity is! HoldScrollActivity;
  }

  _animateTo(int valueToSelect, {int durationMillis = 200}) {
    double targetExtent = (valueToSelect - minValue) * itemExtent;
    scrollController.animateTo(
      targetExtent,
      duration: new Duration(milliseconds: durationMillis),
      curve: Curves.decelerate,
    );
  }

  int _offsetToMiddleIndex(double offset) => (offset + width / 2) ~/ itemExtent;

  int _offsetToMiddleValue(double offset) {
    int indexOfMiddleElement = _offsetToMiddleIndex(offset);
    int middleValue = _indexToValue(indexOfMiddleElement);
    middleValue = math.max(minValue, math.min(maxValue, middleValue));

    return middleValue;
  }

  bool _onNotification(Notification notification) {
    if (notification is ScrollNotification) {
      int middleValue = _offsetToMiddleValue(notification.metrics.pixels);
      if (_userStoppedScrolling(notification)) {
        _animateTo(middleValue);
      }

      if (middleValue != value && this.onChange != null) {
        this.onChange(middleValue);
      }
    }

    return true;
  }
}


lib/src/weight_slider_background.dart
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

class WeightBackground extends StatelessWidget {
  final Widget? child;
  final double? height;

  const WeightBackground({Key? key, this.child, this.height}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: Alignment.bottomCenter,
      children: <Widget>[
        Container(
          height: this.height,
          decoration: BoxDecoration(
            color: Color.fromRGBO(244, 244, 244, 1.0),
            borderRadius: new BorderRadius.circular(50.0),
          ),
          child: child,
        ),
        SvgPicture.asset(
          'images/arrow.svg',
          package: 'weight_slider',
          color: Theme.of(context).primaryColor,
        ),
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:weight_slider/weight_slider.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(
        title: 'Demo',
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int weight = 70;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Padding(
          padding: EdgeInsets.all(64),
          child: WeightSlider(
            weight: weight,
            minWeight: 40,
            maxWeight: 120,
            onChange: (val) => setState(() => this.weight = val),
            unit: 'kg', // optional
          ),
        ),
      ),
    );
  }
}
